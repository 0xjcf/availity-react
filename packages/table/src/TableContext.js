import React, { useContext } from 'react';

export const TableContext = React.createContext();
export const useTableContext = () => useContext(TableContext);
