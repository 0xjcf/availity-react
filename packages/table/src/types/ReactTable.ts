import {
  Row as RtRow,
  Cell as RtCell,
  HeaderGroup,
  Column as RtColumn,
  TableInstance as RtTableInstance,
  TableOptions as RtTableOptions,
  UseSortByInstanceProps,
  UseRowSelectInstanceProps,
  TableState,
  UseSortByColumnProps,
  UseTableHeaderGroupProps,
  UseRowSelectOptions,
  UseTableOptions,
  UseSortByOptions,
  UsePaginationOptions,
  UsePaginationInstanceProps,
  UseColumnOrderInstanceProps,
} from 'react-table';

import { TableSort } from './TableSort';

/* eslint-disable-next-line  @typescript-eslint/ban-types */
export type IdType = { id?: string | number } & object;

export type Row<T extends IdType> = RtRow<T> & {
  toggleRowSelected: () => void;
  getToggleRowSelectedProps: () => void;
  original: T;
};

export type Column<T extends IdType> = RtColumn<T>  & {
  className?: string;
  stickyRight?: boolean;
  stickyLeft?: boolean;
  disableClick?: boolean;
  title?: string;
  defaultCanSort?: boolean;
  disableSortBy?: boolean;
  label?: string;
  hidden?: boolean;
  canCustomize?: boolean;
  isVisible?: boolean;
};

export type Cell<T extends IdType> = RtCell<T> & {
  index: number;
  row: Row<T>;
  column: Column<T>;
  original: T;
};

export type ExtendedTableHeader<T extends IdType> = HeaderGroup<T> &
  Column<T> &
  UseSortByColumnProps<T> &
  UseTableHeaderGroupProps<T> & {
    defaultCanSort?: boolean;
    disableSortBy?: boolean;
    isSorted?: boolean;
    isSortedDesc?: boolean;
    toggleSortBy: (isDesc: boolean, isMult: boolean) => void;
  };

export type TableOptions<T extends IdType> = RtTableOptions<T> &
  UseTableOptions<T> & {
    autoResetSelectedRows?: boolean;
    autoResetSortBy?: boolean;
    manualSortBy?: boolean;
    hiddenColumns?: string[];
    initialState?: Partial<TableState<T>> & CurrentTableState;
  } & UseSortByOptions<T> & UsePaginationOptions<T>;

export type TableInstance<T extends IdType> = UseSortByInstanceProps<T> &
  UseRowSelectInstanceProps<T> &
  UseRowSelectOptions<T> &
  RtTableInstance<T> &
  UseColumnOrderInstanceProps<T> &
  UsePaginationInstanceProps<T> & {
    toggleAllPageRowsSelected: (set?: boolean) => void;
    selectedFlatRows: Row<T>[];
    isAllRowsSelected: boolean;
    toggleAllRowsSelected: () => void;
    sortBy?: TableSort;
  } & {
    manualSortBy?: boolean;
  };

export interface CurrentTableState extends TableState {
  sortBy?: TableSort[];
}
