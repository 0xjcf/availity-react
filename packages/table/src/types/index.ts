export type { OnTableClickEvent } from './OnTableClickEvent';
export type {
    OnRowSelectedEvent
} from './OnRowSelectedEvent';

export type {
    Row,
    Column,
    Cell,
    ExtendedTableHeader,
    CurrentTableState,
    TableInstance
} from './ReactTable';

export type {
    PrimaryTableAction,
    SingleTableAction,
    TableAction,
    BulkTableAction
} from './TableActions';

export type {
    TableSortConfig
} from './TableSortConfig';

export type {
    TableSortOption
} from './TableSortOption';