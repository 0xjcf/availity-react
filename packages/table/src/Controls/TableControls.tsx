import React, { cloneElement, isValidElement, useEffect, useState } from 'react';
import { Form } from 'reactstrap';
import { useTableContext } from '../TableContext';

type Props = {
  id?: string;
  disabled?: boolean;
  children?: React.ReactNode[] | React.ReactNode;
} & React.HTMLAttributes<HTMLElement>;

const TableControls = ({ id, disabled, children, ...rest }: Props): JSX.Element | null => {
  const [isDisabled, setIsDisabled] = useState(disabled || false);
  
  const { instance } = useTableContext();
  const [isReady, setIsReady] = useState(!!instance);

  useEffect(() => {
    if (instance) {
      setIsReady(true);
    }
  }, [instance]);

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const childElements = Array.isArray(children) ? children : [children];
  if (!isReady) {
    return null;
  }
  return (
    <div id={id} className="table-controls" {...rest}>
      <Form inline>
        {childElements?.map((child: React.ReactNode, index: number) => {
          if (!isValidElement(child)) {
            return null;
          }

          return cloneElement(child, { disabled: isDisabled, key: `table_controls_${index}` });
        })}
      </Form>
    </div>
  );
};

export default TableControls;
