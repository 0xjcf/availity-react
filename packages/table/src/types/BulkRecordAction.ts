import { DropdownItemProps } from 'reactstrap';
import { IdType, Row } from './ReactTable';
import { TableAction } from './TableAction';

export interface BulkRecordAction<T extends IdType> extends TableAction {
  onClick?: (records?: T[], rows?: Row<T>[]) => void;
  isVisible?: (records?: T[], rows?: Row<T>[]) => boolean;
  displayText?:
    | string
    | React.ReactChild
    | React.ElementType
    | ((records: T[], rows?: Row<T>[]) => string | React.ReactChild | React.ElementType);
  dropdownItemProps?: DropdownItemProps;
}
