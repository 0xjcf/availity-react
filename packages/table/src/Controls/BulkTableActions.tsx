import React, { useEffect, useState } from 'react';
import { Badge, Button, ButtonDropdown, ButtonGroup, DropdownItem, DropdownMenu, DropdownToggle } from 'reactstrap';
import { useTableContext } from '../TableContext';
import { TableInstance } from '../types/ReactTable';
import { BulkTableAction } from '../types/TableActions';
import { TableRecord } from '../types/TableRecord';

type Props = {
    id?: string;
    disabled?: boolean;
    recordName: string;
    bulkActions: BulkTableAction[];
    color?: string;
    onRecordsSelected?: (records: TableRecord[]) => void;
} & React.HTMLAttributes<HTMLElement>;

const BulkTableActions = ({ id, disabled, color, recordName, bulkActions, onRecordsSelected }: Props): JSX.Element | null => {
    const { instance } = useTableContext();

    const { selectedFlatRows: selectedRows, isAllRowsSelected, toggleAllRowsSelected } = instance as TableInstance;

    const [isSelectionDropdownOpen, setIsSelectionDropdownOpen] = useState<boolean>(false);
    const [numberOfSelectedRows, setNumberOfSelectedRows] = useState<number>(0);
    const [selectionButtonText, setSelectionButtonText] = useState('Select');
    const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);

    useEffect(() => {
        setNumberOfSelectedRows(selectedRows?.length);
        if (isAllRowsSelected) {
            setSelectionButtonText('Deselect');
        } else {
            setSelectionButtonText('Select');
        }

        if (onRecordsSelected) {
            const records = selectedRows.map(row =>  row.original);
            onRecordsSelected(records);
        }

        setIsDisabled(disabled || selectedRows?.length === 0);
    }, [selectedRows, disabled, numberOfSelectedRows]);

    const toggleSelectionDropdown = () => setIsSelectionDropdownOpen(!isSelectionDropdownOpen);

    const handleToggleSelectAll = () => {
        if (toggleAllRowsSelected) {
            toggleAllRowsSelected();
        }
    }

    return (
        <ButtonGroup id={id} disabled={isDisabled} className="btn-group">
            <Button
                disabled={isDisabled}
                onClick={() => { handleToggleSelectAll() }}
                color={color}>
                <Badge>{numberOfSelectedRows}</Badge> {selectionButtonText} All {recordName}
            </Button>
            <ButtonDropdown disabled={isDisabled} isOpen={isSelectionDropdownOpen} toggle={toggleSelectionDropdown}>
                <DropdownToggle disabled={isDisabled} color={color} caret />
                <DropdownMenu color={color}>
                    {bulkActions.map((action) => {
                        const isVisible = action.isVisible ? action.isVisible() : true;
                        const setProps = () => {
                            if (!action.onClick) {
                                return null;
                            }
                            const clickEvent = action.onClick;
                            return { onClick: () => clickEvent(selectedRows.map(row =>  row.original)) };
                        }

                        if (isVisible) {
                            return <DropdownItem key={action.displayText} {...setProps()}>{action.displayText}</DropdownItem>
                        }
                        return null;
                    }
                    )}
                </DropdownMenu>
            </ButtonDropdown>
        </ButtonGroup>
    );
}

export default BulkTableActions;