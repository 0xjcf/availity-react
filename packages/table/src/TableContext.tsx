import React, { useContext } from 'react';
import { TableSortOption } from './types';

export type AvTableContext = {
  scrollable?: boolean;
  setScrollable?: React.Dispatch<React.SetStateAction<boolean>>;
  
  sortable?: boolean;
  setSortable?: React.Dispatch<React.SetStateAction<boolean>>;

  sortableColumns?: TableSortOption[];
  setSortableColumns?: React.Dispatch<React.SetStateAction<TableSortOption[]>>;

  selectable?: boolean;
  setSelectable?: React.Dispatch<React.SetStateAction<boolean>>;

  hasCustomizableColumns?: boolean;
  isCustomizingColumns?: boolean;
  setIsCustomizingColumns?: React.Dispatch<React.SetStateAction<boolean>>;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  instance?: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setInstance?: React.Dispatch<React.SetStateAction<any>>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  selectedRows: any[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setSelectedRows?: React.Dispatch<React.SetStateAction<any[]>>;
};

export const TableContext = React.createContext<AvTableContext>({
  scrollable: false,
  sortable: false,
  selectedRows: []
});

export const useTableContext = (): AvTableContext => useContext(TableContext);
