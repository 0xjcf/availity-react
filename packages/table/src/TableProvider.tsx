import React, { useEffect, useState } from 'react';
import { useRowSelect, useSortBy, useTable, Hooks, UseRowSelectInstanceProps, Column as RtColumn } from 'react-table';
import filter from 'lodash/filter';
import IndeterminateCheckbox from './IndeterminateCheckbox';
import { Cell, Column, CurrentTableState, IdType, TableInstance, TableOptions } from './types/ReactTable';
import { TableSortOption } from './types/TableSortOption';
import { TableContext } from './TableContext';

export type TableProviderProps<T extends IdType> = {
  additionalContent?: React.ElementType;
  additionalContentProps?: Record<string, string | number | boolean | undefined | null>;

  children: React.ReactChild | React.ReactChild[];
  columns: Column<T>[];
  data: T[];
  scrollable?: boolean;
  selectable?: boolean;
  sortable?: boolean;
  initialState?: Partial<CurrentTableState>;
} & TableOptions<T>;

const TableProvider = <T extends IdType>({
  additionalContent: AdditionalContent,
  additionalContentProps,

  columns,
  data,
  selectable,
  scrollable,
  sortable,
  children,
  ...rest
}: TableProviderProps<T>): JSX.Element => {
  let selectionColumn: Column<T>;
  const [isScrollable, setScrollable] = useState<boolean | undefined>(scrollable);

  const getSortableColumns = (): TableSortOption[] =>
    filter(columns, (column) => !column.disableSortBy && column.defaultCanSort).map((column) => {
      const col = column as Column<T>;
      return { value: col.accessor as string, label: col.Header as string };
    });

  const cols = columns as RtColumn<T>[];

  const tableInstance = useTable<T>(
    {
      ...rest,
      columns: cols,
      data,
    } as TableOptions<T>,
    useSortBy,
    useRowSelect,
    (hooks: Hooks<T>) => {
      selectionColumn = {
        id: 'selection',
        title: 'Select record(s)',
        className: 'fixed-width-selection',
        defaultCanSort: false,
        disableSortBy: true,
        disableClick: true,
        Header: ({ getToggleAllRowsSelectedProps }: UseRowSelectInstanceProps<T>) => (
          <div className="text-center">
            <IndeterminateCheckbox
              data-testid="table_header_select_all"
              aria-label="Select all records"
              {...getToggleAllRowsSelectedProps()}
            />
          </div>
        ),
        Cell: ({ row: { getToggleRowSelectedProps, index } }: Cell<T>) => (
          <div className="text-center">
            <IndeterminateCheckbox
              data-testid={`table_header_select_row_${index}`}
              aria-label="Select record"
              {...getToggleRowSelectedProps()}
            />
          </div>
        ),
      };

      hooks.visibleColumns.push((columns: Column<T>[]) => [selectionColumn, ...columns]);
    }
  ) as TableInstance<T>;

  const { toggleHideColumn } = tableInstance;

  useEffect(() => {
    toggleHideColumn('selection', !selectable);
  }, [selectable, toggleHideColumn]);

  return (
    <TableContext.Provider
      value={{
        AdditionalContent,
        additionalContentProps,
        selectable,
        scrollable: isScrollable as boolean | undefined,
        setScrollable,
        sortable,
        sortableColumns: getSortableColumns(),
        instance: tableInstance,
      }}
    >
      {children}
    </TableContext.Provider>
  );
};

export default TableProvider;
