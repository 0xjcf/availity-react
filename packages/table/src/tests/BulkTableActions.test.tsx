import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import every from 'lodash/every';
import basicData from './data/basicData.json';
import TableControls from '../Controls/TableControls';
import BulkTableActions from '../Controls/BulkTableActions';
import TableProvider from '../TableProvider';
import { Column } from '../types/ReactTable';
import { TableContext } from '../TableContext';

const basicColumns = [
  {
    Header: 'First Name',
    accessor: 'first_name',
    defaultCanSort: true,
    disableSortBy: false,
  },
  {
    Header: 'Last Name',
    accessor: 'last_name',
    defaultCanSort: false,
    disableSortBy: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
  },
] as Column<Record<string, string | boolean | number>>[];

const bulkActions = [
  {
    id: 'action1',
    displayText: 'Action 1',
    isVisible: (records?: Record<string, string | boolean | number>[]) => every(records, { isActive: true }),
    onClick: jest.fn(),
  },
  {
    id: 'action2',
    displayText: 'Action 2',
    onClick: jest.fn(),
  },
];

describe('BulkTableActions', () => {
  test('should render default bulk actions', async () => {
    const { container, getByTestId } = render(
      <TableProvider data={basicData} columns={basicColumns}>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('0 Select All Records');
  });

  test('should render bulk actions with record name', async () => {
    const { getByTestId } = render(
      <TableProvider data={basicData} columns={basicColumns}>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} recordName="MyRecords" />
        </TableControls>
      </TableProvider>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('0 Select All MyRecords');
  });

  test('should show number of selected records', async () => {
    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
    };

    const { getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
        }}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </TableContext.Provider>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('1 Select All Records');
  });

  test('should call toggleAllRowsSelected when clicking button', async () => {
    const toggleAllRowsSelected = jest.fn();

    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
      toggleAllRowsSelected,
    };

    const { getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
        }}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </TableContext.Provider>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    fireEvent.click(btnSelectDeselect);

    await waitFor(() => {
      expect(toggleAllRowsSelected).toHaveBeenCalledTimes(1);
    });
  });

  test('should call bulk action when clicking event', async () => {
    const tableInstanceMock = {
      selectedFlatRows: [basicData[0]],
      data: basicData,
      columns: basicColumns,
    };

    const { getByTestId } = render(
      <TableContext.Provider
        value={{
          instance: tableInstanceMock,
        }}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </TableContext.Provider>
    );

    const btnToggle = await waitFor(() => getByTestId('bulk_actions_toggle'));
    expect(btnToggle).not.toBeNull();
    fireEvent.click(btnToggle);

    const actionToDo = bulkActions[0];

    const btnAction = await waitFor(() => getByTestId(`bulk_action_${actionToDo.id}`));
    expect(btnAction).not.toBeNull();

    fireEvent.click(btnAction);

    await waitFor(() => {
      expect(actionToDo.onClick).toHaveBeenCalledTimes(1);
    });
  });
});
