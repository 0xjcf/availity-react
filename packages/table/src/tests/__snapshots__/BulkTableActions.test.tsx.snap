// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkTableActions should render default bulk actions 1`] = `
<div>
  <div
    class="table-controls"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        role="group"
      >
        <button
          class="btn btn-secondary"
          data-testid="select_deselect_all_records"
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
  <table
    class="av-grid table"
    role="table"
  >
    <thead
      class="av-grid-row-header"
      id="table_header"
    >
      <tr
        data-testid="table_header_row_0"
        id="table_header_row_0"
        role="row"
      >
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_0_first_name"
          id="table_header_row_0_cell_0_first_name"
          role="columnheader"
          title="First Name"
        >
          First Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_1_last_name"
          id="table_header_row_0_cell_1_last_name"
          role="columnheader"
          title="Last Name"
        >
          Last Name
        </th>
        <th
          colspan="1"
          data-testid="table_header_row_0_cell_2_email"
          id="table_header_row_0_cell_2_email"
          role="columnheader"
          title="Email"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        class="av-grid-row-even"
        data-testid="table_row_0"
        id="table_row_0"
        role="row"
      >
        <td
          data-testid="table_row_0_cell_0"
          id="table_row_0_cell_0"
          role="cell"
        >
          Barbra
        </td>
        <td
          data-testid="table_row_0_cell_1"
          id="table_row_0_cell_1"
          role="cell"
        >
          Tull
        </td>
        <td
          data-testid="table_row_0_cell_2"
          id="table_row_0_cell_2"
          role="cell"
        >
          btull0@usatoday.com
        </td>
      </tr>
      <tr
        class="av-grid-row-odd"
        data-testid="table_row_1"
        id="table_row_1"
        role="row"
      >
        <td
          data-testid="table_row_1_cell_0"
          id="table_row_1_cell_0"
          role="cell"
        >
          Karlis
        </td>
        <td
          data-testid="table_row_1_cell_1"
          id="table_row_1_cell_1"
          role="cell"
        >
          Yurshev
        </td>
        <td
          data-testid="table_row_1_cell_2"
          id="table_row_1_cell_2"
          role="cell"
        >
          kyurshev1@spotify.com
        </td>
      </tr>
      <tr
        class="av-grid-row-even"
        data-testid="table_row_2"
        id="table_row_2"
        role="row"
      >
        <td
          data-testid="table_row_2_cell_0"
          id="table_row_2_cell_0"
          role="cell"
        >
          Bordie
        </td>
        <td
          data-testid="table_row_2_cell_1"
          id="table_row_2_cell_1"
          role="cell"
        >
          Marsy
        </td>
        <td
          data-testid="table_row_2_cell_2"
          id="table_row_2_cell_2"
          role="cell"
        >
          bmarsy2@elegantthemes.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
